### DISCUSSION TITLE
Hints on how to send/receive GLSP Actions and RequestActions from Theia Widget to manipulate diagram on the server.

### DISCUSSION THREAD

**Discussion by ahm-aigilit** [eclipse-glsp/glsp#1352] [Q&A]
Hey, I am fairly new to GLSP and Theia and building a prototype for a graphical model editor. 
Now I would like to interact with the diagram from a custom Theia ReactWidget. 

Do you have any hints or examples on how I would implement sending custom Actions over the existing Websocket connection. 
I would like to include the opened diagram file URI and also (if any) a potentially selected node ID. 
The diagram might get changed on the server and needs to be updated on the client. This is part of the GLSP protocol, so I would like to use that.

I have the custom widget already integrated in Theia. Now I am a bit lost on how to get hold on the GLSP specifics. Are there any examples out there I can look at?

Appreciate your time for some hints.

**Comment by martin-fleck-at** [comment]
Hi @ahm-aigilit, I hope I get your use case right but it seems that you want to send actions from a particular diagram from outside the diagram, is that right? Currently there is no generic mechanism to send an action to an arbitrary diagram widget since the whole GLSP diagram widget is encapsulated in it's own Inversify container. However, you can retrieve the diagram widget from Theia (either by going through all widgets and finding the one that you need or if it is the active one you could use the `getDiagramWidget` [function](https://github.com/eclipse-glsp/glsp-theia-integration/blob/a5c02896d3d5f602227a1df2f354c1c0c6f43a4b/packages/theia-integration/src/browser/diagram/glsp-diagram-widget.ts#L459)). Once you have the actual diagram widget, you can access the context of the diagram through the exposed `diContainer` and specifically to send actions you could access the `actionDispatcher` [property](https://github.com/eclipse-glsp/glsp-theia-integration/blob/a5c02896d3d5f602227a1df2f354c1c0c6f43a4b/packages/theia-integration/src/browser/diagram/glsp-diagram-widget.ts#L413) from the diagram and send actions to the server like that.
